### PR TITLE
fix: stabilize and smooth conversation awareness volume

### PR DESCRIPTION
**Pull request by Ante042**
Conversation Awareness restored full volume on level `0x02`, which is still a speech event. On my AirPods Pro 3, speaking produced repeated volume changes instead of keeping playback quiet until the conversation ended.

Treat levels 1/2 as speaking, ignore intermediate levels, and restore on 6/8/9, matching the [upstream Android handler](https://github.com/librepods-org/librepods/blob/main/android/app/src/main/java/me/kavishdevar/librepods/services/AirPodsService.kt). Fade PipeWire's software gain over 600 ms down and 1000 ms up, preserving the selected volume and SBC-XQ profile. Serialize software-gain writes asynchronously, coalesce superseded animation values, and restore gain on disconnect, control-socket loss, sink replacement, output removal, and graceful shutdown. Failed writes trigger restoration with a bounded retry while retaining the affected sink.

Validation:
- Reproduced the original `01 -> 02` restoration and added a regression that failed with the old classification. On actual AirPods Pro 3, the corrected classification keeps volume down through speech and restores at the end.
- Hardware volume animation was audibly stepped. A direct software-gain experiment was smoother; the final cancellable animation is installed, with its listening check still pending.
- Build and all 20 tests pass. The new test covers packet validation, repeated speech events, intermediate gain values, reversal, immediate cancellation, reset, and sink replacement.
- Ran the final helper against a suspended PipeWire null sink: reverse at 200 ms, reverse again at 350 ms, reset at 500 ms; gain remained 1.0 at 1600 ms. Removed the test sink afterward.
- Two independent reviews found no remaining actionable defects after fixing the ramp cancellation and reconnect paths.
- Verified real `pw-cli` name lookup separately from the fake test: wrote 0.008 and 1.0 by sink name, read both back by independently resolved numeric node ID, and ran the unchanged helper against the same temporary sink. It reached 0.008 at 850 ms and restored 1.0. See the [source references and results](https://github.com/thisisgm/omarchy-pods/pull/58#discussion_r3995905314).
- Review regressions: the old implementation fails both the stalled-command responsiveness check and the failed-fade restoration check (actual gain 0.008, expected 1.0). The asynchronous implementation passes, including failed restoration before sink replacement and a delayed old-command failure after a replacement is requested. Real PipeWire readback also confirms 0.008 during speech and 1.0 afterward with the updated helper.

The fade uses ordered scalar writes instead of PipeWire sample ramps, because a suspended sink can retain an unfinished sample ramp that overwrites a later reset. This assumes the sink's independent software gain starts at unity; normal volume stays under the existing PulseAudio controller. ANC/transparency transitions remain controlled by the AirPods and are not claimed to be smoother.

Built on ARM64 with Qt 6.11.2 and PipeWire 1.6.8. Fresh CMake configuration needed the installed OpenGL library and Qt6OpenGL paths passed explicitly on this machine; no build or system configuration changes are included. Physical disconnect during a fade and service termination during a fade have not been exercised.

## Summary by CodeRabbit

- **New Features**
  - Conversation Awareness smoothly lowers media volume while you speak and restores it afterward.
  - Gradual volume transitions make ducking and restoration less abrupt.
  - Conversation volume is restored when AirPods disconnect or the connected audio device changes.
  - Volume handling supports switching between audio outputs without prematurely restoring sound.

- **Bug Fixes**
  - Invalid, incomplete, or unknown Conversation Awareness data is safely ignored.
  - Volume restoration is more reliable after connection interruptions and failed audio updates.
  - Repeated restoration failures are handled without leaving volume changes indefinitely unresolved.

### CODE REVIEW

File: daemon/media/conversationvolume.hpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+private:
+    bool send(double gain)
+    {
+        // Software gain leaves Bluetooth's coarse hardware volume and the user's volume setting alone.
+        const QString parameters = QStringLiteral("{ volume: %1 }").arg(gain, 0, 'f', 6);
+        QProcess command;
+        command.start(QStringLiteral("pw-cli"), {QStringLiteral("set-param"), m_sink,
+                                               QStringLiteral("Props"), parameters});
```
[the remark this answers is not in the document]
Comment (Ante042) [reply daemon/media/conversationvolume.hpp]: I checked this against PipeWire 1.6.8 source and the running audio server. The man page calls the argument `object-id`, but the implementation also accepts node names:

- [`do_set_param()` calls `find_global()`](https://github.com/PipeWire/pipewire/blob/1.6.8/src/tools/pw-cli.c#L1897).
- [`find_global()` matches nonnumeric arguments](https://github.com/PipeWire/pipewire/blob/1.6.8/src/tools/pw-cli.c#L536-L552), including the [class-specific name](https://github.com/PipeWire/pipewire/blob/1.6.8/src/tools/pw-cli.c#L424-L427). For nodes, that key is [`PW_KEY_NODE_NAME`](https://github.com/PipeWire/pipewire/blob/1.6.8/src/tools/pw-cli.c#L1355-L1361).

I also tested with real `pw-cli`, not the fake: created a temporary PulseAudio null sink named `conversation_name_review`, located its numeric PipeWire ID independently with `pw-dump`, wrote gain by sink name, and read it back by numeric ID. Both commands exited 0:

```text
requested=0.008 actual=0.008
requested=1.0 actual=1.0
```

Then I ran the unchanged `ConversationVolume` from this PR against that name: readback was 0.008 at 850 ms and 1.0 after restoration. The temporary sink was removed afterward. An independent source review reached the same conclusion.

I agree the fake test alone cannot establish runtime compatibility. This real readback does, and the numeric-only premise is false for the installed version. Could you reassess this finding against that evidence?

**Comment by Ante042** [comment]
@codex review

Please reassess the current commit, including the source references and real PipeWire readback in my reply to the node-ID finding. No production code was changed because the runtime test confirms name-based lookup works. Report any remaining actionable problems.

File: daemon/media/conversationvolume.hpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    explicit ConversationVolume(QObject *parent = nullptr) : QObject(parent)
+    {
+        m_fade.setEasingCurve(QEasingCurve::InOutSine);
+        connect(&m_fade, &QVariantAnimation::valueChanged, this, [this](const QVariant &value) {
+            if (m_fade.state() != QAbstractAnimation::Running)
+                return;
+            if (send(value.toDouble()))
+                m_gain = value.toDouble();
+            else
+                m_fade.stop();
```
[the remark this answers is not in the document]
Comment (Ante042) [reply daemon/media/conversationvolume.hpp]: Fixed in 54daf2f. A failed fade now stops pending animation values and actively restores unity, with one retry if restoration fails. The sink remains tracked until restoration succeeds or the sink is confirmed gone. The regression with a failing fade-up left gain at 0.008 on the previous implementation; it now reaches 1.0 without another conversation packet.

File: daemon/media/conversationvolume.hpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        constexpr int fadeUpMs = 1000;
+        m_fade.setDuration(speaking ? fadeDownMs : fadeUpMs);
+        m_fade.start();
+    }
+
+    void reset()
+    {
+        m_fade.stop();
+        if (!m_sink.isEmpty())
+            send(1.0);
```
[the remark this answers is not in the document]
Comment (Ante042) [reply daemon/media/conversationvolume.hpp]: Fixed in 54daf2f. Failed restoration keeps the old sink and dirty gain state, retries once, and prevents replacement-sink writes from discarding that state. A later lifecycle request can retry after both attempts fail; confirmed sink disappearance releases the old target. The fail-once restoration test verifies both attempts occur before writes to the replacement sink.

File: daemon/media/conversationvolume.hpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+private:
+    bool send(double gain)
+    {
+        // Software gain leaves Bluetooth's coarse hardware volume and the user's volume setting alone.
+        const QString parameters = QStringLiteral("{ volume: %1 }").arg(gain, 0, 'f', 6);
+        QProcess command;
+        command.start(QStringLiteral("pw-cli"), {QStringLiteral("set-param"), m_sink,
+                                               QStringLiteral("Props"), parameters});
+        constexpr int commandTimeoutMs = 2000;
+        const bool finished = command.waitForFinished(commandTimeoutMs);
```
[the remark this answers is not in the document]
Comment (Ante042) [reply daemon/media/conversationvolume.hpp]: Fixed in 54daf2f. One asynchronous QProcess serializes writes and coalesces pending fade values. Reset waits in that sequence, so an old write cannot overtake unity. A stalled fake command no longer blocks an independent 30 ms timer; the same check failed with the synchronous implementation. Bounded synchronous cleanup is limited to destruction, when the event loop is leaving.

File: daemon/main.cpp [hunk 5]
```diff
@@ -825,6 +825,7 @@ private slots:
     void onDeviceDisconnected(const QBluetoothAddress &address)
     {
         LOG_INFO("Device disconnected: " << address.toString());
+        mediaController->resetConversationVolume();
```
[the remark this answers is not in the document]
Comment (Ante042) [reply daemon/main.cpp:828]: Fixed in 54daf2f. The owned control-socket loss path now resets conversation volume immediately after its ownership guard, including recovery while the A2DP sink survives.

File: daemon/media/mediacontroller.cpp [hunk 6]
```diff
@@ -149,55 +153,21 @@ bool MediaController::isActiveOutputDeviceAirPods() {
...
+    LOG_INFO("Conversational awareness level: " << static_cast<quint8>(data[9]));
+    const auto speaking = AirPodsPackets::ConversationalAwareness::parseSpeaking(data);
+    if (!speaking.has_value())
+        return;
+    if (!isActiveOutputDeviceAirPods()) {
+        m_conversationVolume->reset();
+        return;
     }
+
+    m_conversationVolume->setSpeaking(speaking.value(), m_pulseAudio->getDefaultSink());
```
[the remark this answers is not in the document]
Comment (Ante042) [reply daemon/media/mediacontroller.cpp]: Fixed in 54daf2f. Conversation handling reads one default-sink snapshot, validates it with the existing sourceNamesAddress helper, and passes that exact value to the fade. The shared active-output check now uses the same helper, whose existing regression tests reject empty sink names and addresses.

**Comment by Ante042** [comment]
@codex review

Please review 54daf2f. Failed fades now restore unity, restoration retains its sink and retries once, writes are asynchronous and serialized, and control-link loss plus sink validation are covered. The new regressions fail on the previous implementation and pass now. The earlier node-name finding has source and real-server readback evidence in its thread.

File: daemon/media/conversationvolume.hpp [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+    void setSpeaking(bool speaking, const QString &sink)
+    {
+        if (sink != m_requestedSink || speaking != m_speaking) {
+            m_fade.stop();
+            m_pendingWrite = false;
+        }
+        m_requestedSink = sink;
+        m_speaking = speaking;
+        m_restoreFailures = 0;
```
[the remark this answers is not in the document]
Comment (Ante042) [reply daemon/media/conversationvolume.hpp]: Confirmed and fixed in 5f27444. `setSpeaking()` and `reset()` now preserve the dirty sink’s retry budget. It resets only after successful restoration to unity, including a normal fade, or confirmed sink disappearance.

The regression failed before the fix with 3 attempts instead of 2 and now passes. It covers repeated events during and after retry exhaustion, plus a fresh budget after recovery. Focused Qt tests pass; hardware validation was not performed.
